test/storage/checkpoint_test.cpp

```diff
@@ -0,0 +1,194 @@
+
+  transaction::TransactionContext *txn = txn_manager->BeginTransaction();
+  manager.StartCheckpoint(txn);
+  manager.Checkpoint(*table, tested.GetLayout());
+  manager.EndCheckpoint();
+  txn_manager->Commit(txn, StorageTestUtil::EmptyCallback, nullptr);
+  unlink(manager.GetCheckpointFilePath(txn).c_str());
+  delete txn;
+}
+
+// NOLINTNEXTLINE
+TEST_F(CheckpointTests, SimpleCheckpointWithVarlen) {
+  const uint32_t num_rows = 100;
+  const uint32_t num_columns = 3;
+  int magic_seed = 13523;
+
+  auto tested = RandomSqlTableTestObject();
+  std::default_random_engine random_generator(magic_seed);
+  tested.GenerateRandomColumns(num_columns, true, &random_generator);
+  tested.Create();
+  tested.InsertRandomRows(num_rows, 0.2, &random_generator);
+
+  storage::CheckpointManager manager(CHECKPOINT_FILE_PREFIX);
+  storage::SqlTable *table = tested.GetTable();
+  transaction::TransactionManager *txn_manager = tested.GetTxnManager();
+
+  transaction::TransactionContext *txn = txn_manager->BeginTransaction();
+  manager.StartCheckpoint(txn);
+  manager.Checkpoint(*table, tested.GetLayout());
+  manager.EndCheckpoint();
+  txn_manager->Commit(txn, StorageTestUtil::EmptyCallback, nullptr);
+  unlink(manager.GetCheckpointFilePath(txn).c_str());
+  delete txn;
+}
+
+// NOLINTNEXTLINE
+TEST_F(CheckpointTests, SimpleCheckpointRecoveryNoVarlen) {
+  const uint32_t num_rows = 10;
+  const uint32_t num_columns = 3;
+  int magic_seed = 13523;
+  // initialize test
+  auto tested = RandomSqlTableTestObject();
+  std::default_random_engine random_generator(magic_seed);
+  tested.GenerateRandomColumns(num_columns, false, &random_generator);
+  tested.Create();
+  tested.InsertRandomRows(num_rows, 0.2, &random_generator);
+
+  storage::CheckpointManager manager(CHECKPOINT_FILE_PREFIX);
+  storage::SqlTable *table = tested.GetTable();
+  storage::BlockLayout layout = tested.GetLayout();
+  catalog::Schema *schema = tested.GetSchema();
+  transaction::TransactionManager *txn_manager = tested.GetTxnManager();
+
+  // checkpoint
+  transaction::TransactionContext *checkpoint_txn = txn_manager->BeginTransaction();
+  manager.StartCheckpoint(checkpoint_txn);
+  manager.Checkpoint(*table, layout);
+  manager.EndCheckpoint();
+  txn_manager->Commit(checkpoint_txn, StorageTestUtil::EmptyCallback, nullptr);
+  std::string checkpoint_path = manager.GetCheckpointFilePath(checkpoint_txn);
+  // read first run
+  transaction::TransactionContext *scan_txn = txn_manager->BeginTransaction();
+  std::vector<std::string> original_rows;
+  tested.PrintAllRows(scan_txn, table, &layout, &original_rows);
+  txn_manager->Commit(scan_txn, StorageTestUtil::EmptyCallback, nullptr);
+  // recovery to another table
+  transaction::TransactionContext *recovery_txn = txn_manager->BeginTransaction();
+  storage::BlockStore block_store_{10000, 10000};
+  storage::SqlTable *recovered_table = new storage::SqlTable(&block_store_, *schema, catalog::table_oid_t(1));
```

Comment:
In an actual recovery where will you get the schema from?

Reply:
Normally, we should get the schema of a table from the catalog, but the catalog is not yet merged.

Another way to get the schema is to get it from the SqlTable itself. Do you consider adding an API for retrieving the schema from a SqlTable reasonable? Especially with the multi-version schema you guys are working on.